src/linear_algebra/affine_space.lean

```diff
@@ -395,75 +451,104 @@ begin
   split,
   { intro hp,
     rw ←vsub_vadd V p hn.some,
-    refine s2.add _ _ hq2 _,
+    refine direction_vadd_mem _ hq2,
     rw ←hd,
-    exact s1.sub _ _ hp hq1 },
+    exact vsub_mem_direction hp hq1 },
   { intro hp,
     rw ←vsub_vadd V p hn.some,
-    refine s1.add _ _ hq1 _,
+    refine direction_vadd_mem _ hq1,
     rw hd,
-    exact s2.sub _ _ hp hq2 }
+    exact vsub_mem_direction hp hq2 }
 end
 
 /-- Construct an affine subspace from a point and a direction. -/
 def mk_of_point_of_direction (p : P) (direction : submodule k V) : affine_subspace k V P :=
 { carrier := {q | ∃ v ∈ direction, q = v +ᵥ p},
-  direction := direction,
-  nonempty := ⟨p, ⟨0, ⟨direction.zero_mem, (add_action.zero_vadd _ _).symm⟩⟩⟩,
-  add := λ p2 v hp2 hv, begin
-    rcases hp2 with ⟨v2, hv2, hp2⟩,
-    use [v + v2, direction.add_mem hv hv2],
-    rw [←add_action.vadd_assoc, hp2]
-  end,
-  sub := λ p1 p2 hp1 hp2, begin
+  smul_vsub_vadd_mem := λ c p1 p2 p3 hp1 hp2 hp3, begin
     rcases hp1 with ⟨v1, hv1, hp1⟩,
     rcases hp2 with ⟨v2, hv2, hp2⟩,
-    rw [hp1, hp2, vadd_vsub_vadd_cancel_right],
-    exact direction.sub_mem hv1 hv2
+    rcases hp3 with ⟨v3, hv3, hp3⟩,
+    use [c • (v1 - v2) + v3,
+         direction.add_mem (direction.smul_mem c (direction.sub_mem hv1 hv2)) hv3],
+    simp [hp1, hp2, hp3, vadd_assoc]
   end }
 
-/-- The direction of an affine space constructed from a point and a
-direction. -/
-@[simp] lemma direction_mk_of_point_of_direction (p : P) (direction : submodule k V) :
-  (mk_of_point_of_direction p direction).direction = direction :=
-rfl
-
 /-- An affine space constructed from a point and a direction contains
 that point. -/
 lemma mem_mk_of_point_of_direction (p : P) (direction : submodule k V) :
   p ∈ mk_of_point_of_direction p direction :=
 ⟨0, ⟨direction.zero_mem, (add_action.zero_vadd _ _).symm⟩⟩
 
+/-- An affine space constructed from a point and a direction contains
+the result of adding a vector in that direction to that point. -/
+lemma vadd_mem_mk_of_point_of_direction {v : V} (p : P) {direction : submodule k V}
```

Comment:
I would be inclined to rename `mk_of_point_of_direction` to `mk'`. It's less explicit, but currently the `of`s are actually confusing because they are "nested" in a subname.

Reply:
`mem_mk_of_point_of_direction` would become `self_mem_mk'`

Reply:
Renamed as suggested.